src/script/conversation/ConversationVerificationStateHandler/shared/conversation/index.ts

```diff
@@ -0,0 +1,137 @@
+
+export const getActiveConversationsWithUsers = ({
+  conversationState,
+  userIds,
+  userState,
+}: GetActiveConversationsWithUsersParams): GetActiveConversationsWithUsersResult => {
+  return conversationState
+    .filteredConversations()
+    .map((conversationEntity: Conversation) => {
+      if (!conversationEntity.removed_from_conversation()) {
+        const userIdsInConversation = conversationEntity.participating_user_ids().concat(userState.self().qualifiedId);
+        const matchingUserIds = userIdsInConversation.filter(userIdInConversation =>
+          userIds.find(userId => matchQualifiedIds(userId, userIdInConversation)),
+        );
+
+        if (!!matchingUserIds.length) {
+          return {conversationEntity, userIds: matchingUserIds};
+        }
+      }
+      return undefined;
+    })
+    .filter(activeConversationInfo => !!activeConversationInfo) as {
+    conversationEntity: Conversation;
+    userIds: QualifiedId[];
+  }[];
+};
+
+interface GetConversationByGroupIdParams {
+  conversationState: ConversationState;
+  groupId: string;
+}
+export const getConversationByGroupId = ({
+  conversationState,
+  groupId,
+}: GetConversationByGroupIdParams): Conversation | undefined => {
+  return conversationState.filteredConversations().find(conversation => conversation.groupId === groupId);
+};
+
+/**
+ * Check whether to degrade conversation and set corresponding state.
+ *
+ * @param conversationEntity Conversation entity to evaluate
+ * @param shouldShowDegradationWarning Should a modal warn about the degradation?
+ * @returns `true` if conversation state changed to degraded
+ */
+interface WillChangeToDegradedParams {
+  conversationEntity: Conversation;
+  shouldShowDegradationWarning?: boolean;
+  logger: Logger;
+}
+export const willChangeToDegraded = ({
+  conversationEntity,
+  logger,
+  shouldShowDegradationWarning = true,
+}: WillChangeToDegradedParams): boolean => {
+  const state = conversationEntity.verification_state();
+  const isDegraded = state === ConversationVerificationState.DEGRADED;
+
+  if (isDegraded) {
+    return false;
+  }
+
+  // Explicit Boolean check to prevent state changes on undefined
+  const isStateVerified = state === ConversationVerificationState.VERIFIED;
+  const isConversationUnverified = conversationEntity.is_verified() === false;
+  if (isStateVerified && isConversationUnverified) {
+    conversationEntity.verification_state(
+      shouldShowDegradationWarning ? ConversationVerificationState.DEGRADED : ConversationVerificationState.UNVERIFIED,
+    );
+    logger.log(`Verification of conversation '${conversationEntity.id}' changed to degraded`);
+    return true;
+  }
+
+  return false;
+};
+
+/**
+ * Check whether to verify conversation and set corresponding state
+ *
+ * @param conversationEntity Conversation entity to evaluate
+ * @returns `true` if conversation state changed to verified
+ */
+interface WillChangeToVerifiedParams {
```

Comment:
can these two functions  (willChangeToDegraded, willChangeToVerified) reuse a base function and return differently based on a param? seems they are inverse of each other.

Reply:
We would kill readability with that approach.
They do the opposite of each other, and in my head, this means they should be distinct functions.

Reply:
Well the logic is not entirely similar (since the degraded state could be either `DEGRADED` or `UNVERIFIED`) so it might, indeed, not make sense to try to factor those. It's sad that it's the only diff, otherwise it would actually have made a lot of sense to have a generic function that check if a conversation would go from one state to the other.